unittests/AST/ASTImporterTest.cpp

```diff
@@ -3385,6 +3404,38 @@ TEST_P(ImportFriendClasses, ImportOfRepeatedFriendDecl) {
   EXPECT_EQ(ToFriend2, ToImportedFriend2);
 }
 
+TEST_P(ImportFriendClasses, UndeclaredFriendClassShouldNotBeVisible) {
+  Decl *FromTu = getTuDecl("class X { friend class Y; };", Lang_CXX, "from.cc");
+  auto *FromX = FirstDeclMatcher<CXXRecordDecl>().match(
+      FromTu, cxxRecordDecl(hasName("X")));
+  auto *FromFriend = FirstDeclMatcher<FriendDecl>().match(FromTu, friendDecl());
+  RecordDecl *FromRecordOfFriend =
+      const_cast<RecordDecl *>(getRecordDeclOfFriend(FromFriend));
+
+  ASSERT_EQ(FromRecordOfFriend->getDeclContext(), cast<DeclContext>(FromTu));
+  ASSERT_EQ(FromRecordOfFriend->getLexicalDeclContext(),
+            cast<DeclContext>(FromX));
+  ASSERT_FALSE(
+      FromRecordOfFriend->getDeclContext()->containsDecl(FromRecordOfFriend));
+  ASSERT_FALSE(FromRecordOfFriend->getLexicalDeclContext()->containsDecl(
```

Comment:
Could you please add an ASSERT_TRUE for the normal lookup (`DeclContext::noload_lookup(DeclarationName Name)`) and an EXPECT for the "to" context later.
Normal lookup should find this, even if `containsDecl` is false. This is because the `DeclContext` has a twofold structure:
1) List of declarations handled by `addDecl` and `ContainsDecl`. This maintains the parent-child relations.
2) A Lookup table, in which the normal C/C++ lookup searches. This controls the C/C++ visibility rules, everything in the lookup table can be found, thus some Decls are in the list of declarations but are not put into the table (e.g. specializations).